File: 2021/day_5/day.rb

def calc(data)
  
end

def calc_2(data)
  m = Hash.new(0)
  data.each { |d| 
    x1, y1, x2, y2 = d.match(/(\d+),(\d+) -> (\d+),(\d+)/).captures.map(&:to_i)
    d = [(x2-x1).abs, (y2-y1).abs].max + 1
    x = x2 <=> x1
    y = y2 <=> y1
    d.times { |i| m[[x1+x*i, y1+y*i]] += 1 }
  }
  m.select { |k, v| v > 1 }.keys.count
end


describe "day 1 1 passes" do
    let(:sample) {[
"0,9 -> 5,9",
"8,0 -> 0,8",
"9,4 -> 3,4",
"2,2 -> 2,1",
"7,0 -> 7,4",
"6,4 -> 2,0",
"0,9 -> 2,9",
"3,4 -> 1,4",
"0,0 -> 8,8",
"5,5 -> 8,2",
]}

  context "first" do
    xit "Processes sample properly" do
      expect(calc sample).to eq(5)
    end

    xit "Processes file" do
      expect(calc File.open("data").readlines).to eq(7436)
    end
  end

  context "second" do
    it "Processes sample properly" do
      expect(calc_2 sample).to eq(12)
    end

    it "Processes file" do
      expect(calc_2 File.open("data").readlines).to eq(21104)
    end
  end
end

File: 2021/day_5/day.py
import unittest
import re
from collections import defaultdict
from numpy import sign

def calc(data):
  d = defaultdict(int)
  for l in data:
    x1, y1, x2, y2 = map(int, re.match("(\d+),(\d+) -> (\d+),(\d+)", l).groups())
    if x1 == x2:
      if y1 > y2:
        y1, y2 = y2, y1
      for i in range(y1,y2+1):
        d[x1,i] += 1
    if y1 == y2:
      if x1 > x2:
        x1, x2 = x2, x1
      for i in range(x1,x2+1):
        d[i, y1] += 1

  return sum([1 for k, v in d.iteritems() if v > 1])

def iterate(line):
  x1, y1, x2, y2 = line
  x = x2 - x1
  y = y2 - y1
  d = max(abs(x), abs(y))
  x = sign(x)
  y = sign(y)
  for i in range(d+1):
    yield x1+x*i, y1+y*i

def calc_2(data):
  d = defaultdict(int)
  for l in data:
    for x, y in iterate(map(int, re.match("(\d+),(\d+) -> (\d+),(\d+)", l).groups())):
      d[x, y] += 1

  print [k for k, v in d.iteritems() if v > 1]
  return sum([1 for k, v in d.iteritems() if v > 1])

class TestDay(unittest.TestCase):
    sample = [
"0,9 -> 5,9",
"8,0 -> 0,8",
"9,4 -> 3,4",
"2,2 -> 2,1",
"7,0 -> 7,4",
"6,4 -> 2,0",
"0,9 -> 2,9",
"3,4 -> 1,4",
"0,0 -> 8,8",
"5,5 -> 8,2",
]

    def test_first_challenge_data(self):
        self.assertEqual(calc(self.sample), 5)

    def test_final_data(self):
        self.assertEqual(calc(open('data').readlines()), 7436)

    def test_second_challenge_sample(self):
        self.assertEqual(calc_2(self.sample), 12)

    def test_final_data_2(self):
        self.assertEqual(calc_2(open('data').readlines()), 21104)

if __name__ == '__main__':
    unittest.main()
